File: week-2/word-pattern/word-pattern.js
/**
 * @param {string} pattern
 * @param {string} str
 * @return {boolean}
 */
var wordPattern = function(pattern, str) {
    if (!pattern || !str) return false;
    
    return createMapping(pattern) === createMapping(str.split(' '));
};

const createMapping = (string) => {
    const hashtable = {};
    let counter = 0;
    let mapping = '';
    for (let i of string) {
        if (!(i in hashtable)) {
            hashtable[i] = counter;
            ++counter;
        }
        mapping += hashtable[i];
    }
    
    return mapping;
}

File: week-2/word-pattern/word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, str: str) -> bool:
        if not pattern or not str: return False
        return self.createMapping(pattern) == self.createMapping(str.split(' '))
    
    def createMapping(self, string):
        count = 0
        mapping = ''
        hashtable = {}
        for i in string:
            if i not in hashtable:
                hashtable[i] = str(count)
                count += 1
            mapping += hashtable[i]
            
        return mapping
